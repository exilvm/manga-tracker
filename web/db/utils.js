const format = require('pg-format');
const {
  NUMERIC_VALUE_OUT_OF_RANGE,
  INVALID_TEXT_REPRESENTATION,
} = require('pg-error-constants');
const dblog = require('debug')('db');

function generateEqualsColumns(o, availableColumns) {
  const cols = new Set(availableColumns);
  const sqlValues = [];
  const columns = [];
  const args = [];

  Object.keys(o)
    .filter(k => o[k] !== undefined && cols.has(k))
    .forEach((k, idx) => {
      sqlValues.push(`%I=$${idx+1}`);
      columns.push(k);
      args.push(o[k]);
    });

  const sqlCols = format(sqlValues.join(','), ...columns);

  return { sqlCols, args };
}
module.exports.generateEqualsColumns = generateEqualsColumns;

function handleError(err, res) {
  if (err.code === INVALID_TEXT_REPRESENTATION) {
    dblog(err.message);
    res.status(400).json({ error: 'Invalid data type given' });
    return;
  }

  if (err.code === NUMERIC_VALUE_OUT_OF_RANGE) {
    res.status(400).json({ error: 'Number value out of range' });
    return;
  }

  console.error(err);
  res.status(500).json({ error: 'Internal server error' });
}
module.exports.handleError = handleError;
