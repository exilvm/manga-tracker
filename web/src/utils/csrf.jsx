import { createContext, useContext } from 'react';

const CSRFContext = createContext(undefined);
export const CSRFProvider = CSRFContext.Provider;

export const useCSRF = () => useContext(CSRFContext);

/**
 * Returns the useCSRF token in the props property with the key _csrf
 * @param req
 * @returns {{props: {_csrf: String}}}
 */
export const csrfProps = ({ req }) => ({
  props: { _csrf: req.csrfToken() },
});

export const csrfHeader = (csrf) => ({
  'X-CSRF-Token': csrf,
});
