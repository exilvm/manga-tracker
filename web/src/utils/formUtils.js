export const showErrorAlways = ({
  meta: { submitError, dirtySinceLastSubmit, error },
}) => !!(((submitError && !dirtySinceLastSubmit) || error));
