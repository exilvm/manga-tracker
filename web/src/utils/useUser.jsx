import { createContext, useContext, useMemo } from 'react';


const UserContext = createContext(undefined);
export const UserProvider = UserContext.Provider;

export const useUser = () => {
  const user = useContext(UserContext);
  return useMemo(() => ({
    user,
    isAuthenticated: !!user,
    isAdmin: user && user.admin,
  }), [user]);
};
