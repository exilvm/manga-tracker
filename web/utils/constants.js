const csrfMissing = 'Missing CSRF token';
const isDev = process.env.NODE_ENV !== 'production';
const isTest = process.env.NODE_ENV === 'test';


module.exports = {
  csrfMissing,
  isDev,
  isTest,
};
