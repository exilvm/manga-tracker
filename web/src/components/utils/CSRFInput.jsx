import { Field } from 'react-final-form';
import { useCSRF } from '../../utils/csrf';

export default function CSRFInput() {
  const csrf = useCSRF();
  return (
    <Field
      name='_csrf'
      component='input'
      type='hidden'
      defaultValue={csrf}
    />
  );
}
