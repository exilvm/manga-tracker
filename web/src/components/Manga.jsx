import React, { useCallback, useState } from 'react';
import {
  Button,
  Container,
  Grid, IconButton,
  Paper,
  Tooltip,
  Typography,
} from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';

import { Edit as EditIcon } from '@material-ui/icons';
import MangaSourceList from './MangaSourceList';
import {
  defaultDateDistanceToNow,
  defaultDateFormat,
  followUnfollow,
} from '../utils/utilities';
import ChapterList from './ChapterList';
import { useUser } from '../utils/useUser';


const useStyles = makeStyles((theme) => ({
  title: {
    width: '75%',
    textAlign: 'left',
    paddingBottom: '10px',
    [theme.breakpoints.down('sm')]: {
      width: '100%',
    },
  },
  titleBar: {
    display: 'flex',
    justifyContent: 'space-between',
  },
  thumbnail: {
    maxWidth: '250px',
    maxHeight: '355px',
    [theme.breakpoints.down('sm')]: {
      maxWidth: '180px',
    },
    [theme.breakpoints.down('xs')]: {
      maxWidth: '125px',
    },
  },
  details: {
    display: 'flex',
  },
  detailText: {
    marginLeft: '5px',
    [theme.breakpoints.down('sm')]: {
      marginLeft: '3px',
    },
  },
  infoTable: {
    marginLeft: '30px',
    marginTop: '3px',
    [theme.breakpoints.down('sm')]: {
      marginLeft: '20px',
    },
    [theme.breakpoints.down('xs')]: {
      marginLeft: '10px',
    },
  },
  sourceList: {
    marginLeft: '35px',
    [theme.breakpoints.down('sm')]: {
      marginLeft: '23px',
    },
    [theme.breakpoints.down('xs')]: {
      marginLeft: '13px',
    },
  },
  followButton: {
    marginTop: theme.spacing(2),
    marginBottom: theme.spacing(2),
  },
  paper: {
    padding: '1em',
    minWidth: '440px',
  },
}));


function Manga(props) {
  const {
    mangaData,
    userFollows = [],
  } = props;

  const { isAuthenticated, user } = useUser();

  const classes = useStyles();
  const [editing, setEditing] = useState(false);
  const startEditing = useCallback(() => setEditing(!editing), [editing]);

  const latestRelease = mangaData.latest_release ?
    new Date(mangaData.latest_release) :
    null;

  const estimatedRelease = new Date(mangaData.estimated_release);

  const mangaChapters = React.useMemo(() => {
    if (!mangaData.chapters) return null;
    const serviceMap = {};
    mangaData.services.forEach(service => { serviceMap[service.service_id] = service.url_format });
    return mangaData.chapters.map(chapter => {
      const newChapter = { ...chapter };
      newChapter.release_date = new Date(chapter.release_date);
      newChapter.url = serviceMap[chapter.service_id].replace('{}', chapter.chapter_url);
      return newChapter;
    });
  }, [mangaData.chapters, mangaData.services]);

  return (
    <Container maxWidth='lg'>
      <Paper className={classes.paper}>
        <div className={classes.titleBar}>
          <Typography className={classes.title} variant='h4'>{mangaData.title}</Typography>
          {user?.admin && (
            <IconButton onClick={startEditing}>
              <EditIcon />
            </IconButton>
          )}
        </div>
        <div className={classes.details}>
          <a href={mangaData.mal} target='_blank' rel='noreferrer noopener'>
            <img
              src={mangaData.cover}
              className={classes.thumbnail}
              alt={mangaData.title}
            />
          </a>
          <Grid
            container
            justify='space-between'
          >
            <table className={classes.infoTable}>
              <tbody>
                <tr>
                  <td><Typography>Latest release:</Typography></td>
                  <td>
                    <Tooltip title={latestRelease ? latestRelease.toUTCString() : 'Unknown'}>
                      <Typography className={classes.detailText}>
                        {latestRelease ?
                          defaultDateFormat(latestRelease) + ' - ' + defaultDateDistanceToNow(latestRelease) :
                          'Unknown'}
                      </Typography>
                    </Tooltip>
                  </td>
                </tr>
                <tr>
                  <td><Typography>Estimated release interval:</Typography></td>
                  <td>
                    <Typography className={classes.detailText}>
                      {(mangaData.release_interval ?
                        `${mangaData.release_interval?.days || 0} days ${mangaData.release_interval?.hours || 0} hours` :
                        'Unknown')}
                    </Typography>
                  </td>
                </tr>
                <tr>
                  <td><Typography>Estimated next release:</Typography></td>
                  <td>
                    <Typography className={classes.detailText}>
                      {defaultDateFormat(estimatedRelease)}
                    </Typography>
                  </td>
                </tr>
                <tr>
                  <td><Typography>Latest chapter:</Typography></td>
                  <td>
                    <Typography className={classes.detailText}>
                      {mangaData.latest_chapter ? mangaData.latest_chapter : 'Unknown'}
                    </Typography>
                  </td>
                </tr>
              </tbody>
            </table>
            <MangaSourceList
              classesProp={[classes.sourceList]}
              items={mangaData.services}
              userFollows={userFollows}
              followUnfollow={(serviceId) => followUnfollow(mangaData.manga_id, serviceId)}
            />
          </Grid>
        </div>
        {isAuthenticated && (
          <Button
            variant='contained'
            color='primary'
            onClick={followUnfollow(mangaData.manga_id, null)}
            className={classes.followButton}
          >
            {userFollows.indexOf(null) < 0 ? 'Follow' : 'Unfollow'}
          </Button>
        )}

        <ChapterList
          chapters={mangaChapters}
          editable={editing}
        />
      </Paper>
    </Container>
  );
}

export default Manga;
