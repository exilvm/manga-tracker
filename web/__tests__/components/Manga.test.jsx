import { createShallow } from '@material-ui/core/test-utils';
import React from 'react';
import Manga from '../../src/components/Manga';
import { emptyFullManga as emptyManga, fullManga as manga } from '../constants';
import { adminUser, mockUTCDates, normalUser, withUser } from '../utils';

describe('Manga page should render correctly', () => {
  mockUTCDates();
  const follows = [1];

  it('should render correctly', () => {
    const wrapper = createShallow()(
      <Manga mangaData={manga} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when logged in', async () => {
    const elem = await withUser(
      normalUser,
      <Manga mangaData={manga} follows={follows} />
    );
    const wrapper = createShallow()(elem).dive();

    expect(wrapper).toMatchSnapshot();
  });

  it('Should render correctly as admin', async () => {
    const elem = await withUser(
      adminUser,
      <Manga mangaData={manga} follows={follows} />
    );
    const wrapper = createShallow()(elem).dive();

    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with little data', () => {
    const wrapper = createShallow()(<Manga mangaData={emptyManga} />);

    expect(wrapper).toMatchSnapshot();
  });
});
