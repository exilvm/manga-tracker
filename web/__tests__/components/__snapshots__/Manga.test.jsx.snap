// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manga page should render correctly Should render correctly as admin 1`] = `
<WithStyles(ForwardRef(Container))
  disableGutters={true}
  maxWidth="lg"
>
  <WithStyles(ForwardRef(Paper))
    className="makeStyles-paper-10"
  >
    <div
      className="makeStyles-titleBar-2"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-title-1"
        variant="h4"
      >
        Dr. STONE
      </WithStyles(ForwardRef(Typography))>
    </div>
    <div
      className="makeStyles-details-5"
    >
      <a
        href="https://myanimelist.net/manga/103897"
        rel="noreferrer noopener"
        target="_blank"
      >
        <img
          alt="Dr. STONE"
          className="makeStyles-thumbnail-4"
          src="https://mangadex.org/images/manga/20882.jpg?1585634146"
        />
      </a>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-rootGrid-12"
        container={true}
        justify="space-between"
      >
        <WithStyles(ForwardRef(Grid))
          className="makeStyles-infoGrid-11"
          container={true}
          direction="column"
        >
          <MangaInfo
            mangaData={
              Object {
                "al": "https://anilist.co/manga/98416",
                "amz": "https://www.amazon.co.jp/gp/product/B075F8JBQ1",
                "ap": "https://www.anime-planet.com/manga/dr-stone",
                "artist": "Boichi",
                "author": "Inagaki Riichiro",
                "bw": "https://bookwalker.jp/series/114645",
                "chapters": Array [
                  Object {
                    "chapter_number": 157,
                    "chapter_url": "1007322",
                    "group": "Shueisha",
                    "release_date": 1593964800000,
                    "service_id": 1,
                    "title": "Z=157: Same Time, Same Place",
                  },
                  Object {
                    "chapter_number": 156,
                    "chapter_url": "938629",
                    "group": "MangaPlus",
                    "release_date": null,
                    "service_id": 2,
                    "title": "Z=156: Two Scientists",
                  },
                  Object {
                    "chapter_number": 156,
                    "chapter_url": "1007024",
                    "group": "Shueisha",
                    "release_date": 1593187200000,
                    "service_id": 1,
                    "title": "Z=156: Two Scientists",
                  },
                ],
                "cover": "https://mangadex.org/images/manga/20882.jpg?1585634146",
                "ebj": "https://www.ebookjapan.jp/ebj/413780/",
                "engtl": "https://www.viz.com/dr-stone",
                "estimated_release": "2020-07-12T16:00:00.000Z",
                "kt": "https://kitsu.io/manga/38860",
                "last_updated": "2020-06-28T08:15:55.170Z",
                "latest_chapter": 157,
                "latest_release": "2020-07-05T16:00:00.000Z",
                "mal": "https://myanimelist.net/manga/103897",
                "manga_id": 1,
                "mu": "https://www.mangaupdates.com/series.html?id=139601",
                "nu": "https://www.novelupdates.com/series/null",
                "raw": "null",
                "release_interval": Object {
                  "days": 7,
                },
                "services": Array [
                  Object {
                    "name": "MANGA Plus",
                    "service_id": 1,
                    "title_id": "100010",
                    "url": "https://mangaplus.shueisha.co.jp/titles/{}",
                    "url_format": "https://mangaplus.shueisha.co.jp/viewer/{}",
                  },
                  Object {
                    "name": "MangaDex",
                    "service_id": 2,
                    "title_id": "20882",
                    "url": "https://mangadex.org/title/{}",
                    "url_format": "https://mangadex.org/chapter/{}",
                  },
                ],
                "status": 0,
                "title": "Dr. STONE",
              }
            }
          />
          <MangaAliases />
        </WithStyles(ForwardRef(Grid))>
        <MangaSourceList
          classesProp={
            Array [
              "makeStyles-sourceList-8",
            ]
          }
          followUnfollow={[Function]}
          items={
            Array [
              Object {
                "name": "MANGA Plus",
                "service_id": 1,
                "title_id": "100010",
                "url": "https://mangaplus.shueisha.co.jp/titles/{}",
                "url_format": "https://mangaplus.shueisha.co.jp/viewer/{}",
              },
              Object {
                "name": "MangaDex",
                "service_id": 2,
                "title_id": "20882",
                "url": "https://mangadex.org/title/{}",
                "url_format": "https://mangadex.org/chapter/{}",
              },
            ]
          }
          userFollows={Array []}
        />
      </WithStyles(ForwardRef(Grid))>
    </div>
    <WithStyles(ForwardRef(Tabs))
      onChange={[Function]}
      value={0}
    >
      <WithStyles(ForwardRef(Tab))
        label="Chapters"
        value={0}
      />
      <WithStyles(ForwardRef(Tab))
        label="Stats"
        value={1}
      />
    </WithStyles(ForwardRef(Tabs))>
    <TabPanelCustom
      index={0}
      noRerenderOnChange={true}
      value={0}
    >
      <ChapterList
        chapters={
          Array [
            Object {
              "chapter_number": 157,
              "chapter_url": "1007322",
              "group": "Shueisha",
              "release_date": 2020-07-05T16:00:00.000Z,
              "service_id": 1,
              "title": "Z=157: Same Time, Same Place",
              "url": "https://mangaplus.shueisha.co.jp/viewer/1007322",
            },
            Object {
              "chapter_number": 156,
              "chapter_url": "938629",
              "group": "MangaPlus",
              "release_date": 1970-01-01T00:00:00.000Z,
              "service_id": 2,
              "title": "Z=156: Two Scientists",
              "url": "https://mangadex.org/chapter/938629",
            },
            Object {
              "chapter_number": 156,
              "chapter_url": "1007024",
              "group": "Shueisha",
              "release_date": 2020-06-26T16:00:00.000Z,
              "service_id": 1,
              "title": "Z=156: Two Scientists",
              "url": "https://mangaplus.shueisha.co.jp/viewer/1007024",
            },
          ]
        }
        editable={false}
        mangaId={1}
        serviceUrlFormats={
          Object {
            "1": "https://mangaplus.shueisha.co.jp/viewer/{}",
            "2": "https://mangadex.org/chapter/{}",
          }
        }
      />
    </TabPanelCustom>
    <TabPanelCustom
      index={1}
      noRerenderOnChange={true}
      value={0}
    >
      <ReleaseHeatmap
        dataRows={Array []}
      />
    </TabPanelCustom>
  </WithStyles(ForwardRef(Paper))>
</WithStyles(ForwardRef(Container))>
`;

exports[`Manga page should render correctly should render correctly 1`] = `
<WithStyles(ForwardRef(Container))
  disableGutters={true}
  maxWidth="lg"
>
  <WithStyles(ForwardRef(Paper))
    className="makeStyles-paper-10"
  >
    <div
      className="makeStyles-titleBar-2"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-title-1"
        variant="h4"
      >
        Dr. STONE
      </WithStyles(ForwardRef(Typography))>
    </div>
    <div
      className="makeStyles-details-5"
    >
      <a
        href="https://myanimelist.net/manga/103897"
        rel="noreferrer noopener"
        target="_blank"
      >
        <img
          alt="Dr. STONE"
          className="makeStyles-thumbnail-4"
          src="https://mangadex.org/images/manga/20882.jpg?1585634146"
        />
      </a>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-rootGrid-12"
        container={true}
        justify="space-between"
      >
        <WithStyles(ForwardRef(Grid))
          className="makeStyles-infoGrid-11"
          container={true}
          direction="column"
        >
          <MangaInfo
            mangaData={
              Object {
                "al": "https://anilist.co/manga/98416",
                "amz": "https://www.amazon.co.jp/gp/product/B075F8JBQ1",
                "ap": "https://www.anime-planet.com/manga/dr-stone",
                "artist": "Boichi",
                "author": "Inagaki Riichiro",
                "bw": "https://bookwalker.jp/series/114645",
                "chapters": Array [
                  Object {
                    "chapter_number": 157,
                    "chapter_url": "1007322",
                    "group": "Shueisha",
                    "release_date": 1593964800000,
                    "service_id": 1,
                    "title": "Z=157: Same Time, Same Place",
                  },
                  Object {
                    "chapter_number": 156,
                    "chapter_url": "938629",
                    "group": "MangaPlus",
                    "release_date": null,
                    "service_id": 2,
                    "title": "Z=156: Two Scientists",
                  },
                  Object {
                    "chapter_number": 156,
                    "chapter_url": "1007024",
                    "group": "Shueisha",
                    "release_date": 1593187200000,
                    "service_id": 1,
                    "title": "Z=156: Two Scientists",
                  },
                ],
                "cover": "https://mangadex.org/images/manga/20882.jpg?1585634146",
                "ebj": "https://www.ebookjapan.jp/ebj/413780/",
                "engtl": "https://www.viz.com/dr-stone",
                "estimated_release": "2020-07-12T16:00:00.000Z",
                "kt": "https://kitsu.io/manga/38860",
                "last_updated": "2020-06-28T08:15:55.170Z",
                "latest_chapter": 157,
                "latest_release": "2020-07-05T16:00:00.000Z",
                "mal": "https://myanimelist.net/manga/103897",
                "manga_id": 1,
                "mu": "https://www.mangaupdates.com/series.html?id=139601",
                "nu": "https://www.novelupdates.com/series/null",
                "raw": "null",
                "release_interval": Object {
                  "days": 7,
                },
                "services": Array [
                  Object {
                    "name": "MANGA Plus",
                    "service_id": 1,
                    "title_id": "100010",
                    "url": "https://mangaplus.shueisha.co.jp/titles/{}",
                    "url_format": "https://mangaplus.shueisha.co.jp/viewer/{}",
                  },
                  Object {
                    "name": "MangaDex",
                    "service_id": 2,
                    "title_id": "20882",
                    "url": "https://mangadex.org/title/{}",
                    "url_format": "https://mangadex.org/chapter/{}",
                  },
                ],
                "status": 0,
                "title": "Dr. STONE",
              }
            }
          />
          <MangaAliases />
        </WithStyles(ForwardRef(Grid))>
        <MangaSourceList
          classesProp={
            Array [
              "makeStyles-sourceList-8",
            ]
          }
          followUnfollow={[Function]}
          items={
            Array [
              Object {
                "name": "MANGA Plus",
                "service_id": 1,
                "title_id": "100010",
                "url": "https://mangaplus.shueisha.co.jp/titles/{}",
                "url_format": "https://mangaplus.shueisha.co.jp/viewer/{}",
              },
              Object {
                "name": "MangaDex",
                "service_id": 2,
                "title_id": "20882",
                "url": "https://mangadex.org/title/{}",
                "url_format": "https://mangadex.org/chapter/{}",
              },
            ]
          }
          userFollows={Array []}
        />
      </WithStyles(ForwardRef(Grid))>
    </div>
    <WithStyles(ForwardRef(Tabs))
      onChange={[Function]}
      value={0}
    >
      <WithStyles(ForwardRef(Tab))
        label="Chapters"
        value={0}
      />
      <WithStyles(ForwardRef(Tab))
        label="Stats"
        value={1}
      />
    </WithStyles(ForwardRef(Tabs))>
    <TabPanelCustom
      index={0}
      noRerenderOnChange={true}
      value={0}
    >
      <ChapterList
        chapters={
          Array [
            Object {
              "chapter_number": 157,
              "chapter_url": "1007322",
              "group": "Shueisha",
              "release_date": 2020-07-05T16:00:00.000Z,
              "service_id": 1,
              "title": "Z=157: Same Time, Same Place",
              "url": "https://mangaplus.shueisha.co.jp/viewer/1007322",
            },
            Object {
              "chapter_number": 156,
              "chapter_url": "938629",
              "group": "MangaPlus",
              "release_date": 1970-01-01T00:00:00.000Z,
              "service_id": 2,
              "title": "Z=156: Two Scientists",
              "url": "https://mangadex.org/chapter/938629",
            },
            Object {
              "chapter_number": 156,
              "chapter_url": "1007024",
              "group": "Shueisha",
              "release_date": 2020-06-26T16:00:00.000Z,
              "service_id": 1,
              "title": "Z=156: Two Scientists",
              "url": "https://mangaplus.shueisha.co.jp/viewer/1007024",
            },
          ]
        }
        editable={false}
        mangaId={1}
        serviceUrlFormats={
          Object {
            "1": "https://mangaplus.shueisha.co.jp/viewer/{}",
            "2": "https://mangadex.org/chapter/{}",
          }
        }
      />
    </TabPanelCustom>
    <TabPanelCustom
      index={1}
      noRerenderOnChange={true}
      value={0}
    >
      <ReleaseHeatmap
        dataRows={Array []}
      />
    </TabPanelCustom>
  </WithStyles(ForwardRef(Paper))>
</WithStyles(ForwardRef(Container))>
`;

exports[`Manga page should render correctly should render correctly when logged in 1`] = `
<WithStyles(ForwardRef(Container))
  disableGutters={true}
  maxWidth="lg"
>
  <WithStyles(ForwardRef(Paper))
    className="makeStyles-paper-10"
  >
    <div
      className="makeStyles-titleBar-2"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-title-1"
        variant="h4"
      >
        Dr. STONE
      </WithStyles(ForwardRef(Typography))>
    </div>
    <div
      className="makeStyles-details-5"
    >
      <a
        href="https://myanimelist.net/manga/103897"
        rel="noreferrer noopener"
        target="_blank"
      >
        <img
          alt="Dr. STONE"
          className="makeStyles-thumbnail-4"
          src="https://mangadex.org/images/manga/20882.jpg?1585634146"
        />
      </a>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-rootGrid-12"
        container={true}
        justify="space-between"
      >
        <WithStyles(ForwardRef(Grid))
          className="makeStyles-infoGrid-11"
          container={true}
          direction="column"
        >
          <MangaInfo
            mangaData={
              Object {
                "al": "https://anilist.co/manga/98416",
                "amz": "https://www.amazon.co.jp/gp/product/B075F8JBQ1",
                "ap": "https://www.anime-planet.com/manga/dr-stone",
                "artist": "Boichi",
                "author": "Inagaki Riichiro",
                "bw": "https://bookwalker.jp/series/114645",
                "chapters": Array [
                  Object {
                    "chapter_number": 157,
                    "chapter_url": "1007322",
                    "group": "Shueisha",
                    "release_date": 1593964800000,
                    "service_id": 1,
                    "title": "Z=157: Same Time, Same Place",
                  },
                  Object {
                    "chapter_number": 156,
                    "chapter_url": "938629",
                    "group": "MangaPlus",
                    "release_date": null,
                    "service_id": 2,
                    "title": "Z=156: Two Scientists",
                  },
                  Object {
                    "chapter_number": 156,
                    "chapter_url": "1007024",
                    "group": "Shueisha",
                    "release_date": 1593187200000,
                    "service_id": 1,
                    "title": "Z=156: Two Scientists",
                  },
                ],
                "cover": "https://mangadex.org/images/manga/20882.jpg?1585634146",
                "ebj": "https://www.ebookjapan.jp/ebj/413780/",
                "engtl": "https://www.viz.com/dr-stone",
                "estimated_release": "2020-07-12T16:00:00.000Z",
                "kt": "https://kitsu.io/manga/38860",
                "last_updated": "2020-06-28T08:15:55.170Z",
                "latest_chapter": 157,
                "latest_release": "2020-07-05T16:00:00.000Z",
                "mal": "https://myanimelist.net/manga/103897",
                "manga_id": 1,
                "mu": "https://www.mangaupdates.com/series.html?id=139601",
                "nu": "https://www.novelupdates.com/series/null",
                "raw": "null",
                "release_interval": Object {
                  "days": 7,
                },
                "services": Array [
                  Object {
                    "name": "MANGA Plus",
                    "service_id": 1,
                    "title_id": "100010",
                    "url": "https://mangaplus.shueisha.co.jp/titles/{}",
                    "url_format": "https://mangaplus.shueisha.co.jp/viewer/{}",
                  },
                  Object {
                    "name": "MangaDex",
                    "service_id": 2,
                    "title_id": "20882",
                    "url": "https://mangadex.org/title/{}",
                    "url_format": "https://mangadex.org/chapter/{}",
                  },
                ],
                "status": 0,
                "title": "Dr. STONE",
              }
            }
          />
          <MangaAliases />
        </WithStyles(ForwardRef(Grid))>
        <MangaSourceList
          classesProp={
            Array [
              "makeStyles-sourceList-8",
            ]
          }
          followUnfollow={[Function]}
          items={
            Array [
              Object {
                "name": "MANGA Plus",
                "service_id": 1,
                "title_id": "100010",
                "url": "https://mangaplus.shueisha.co.jp/titles/{}",
                "url_format": "https://mangaplus.shueisha.co.jp/viewer/{}",
              },
              Object {
                "name": "MangaDex",
                "service_id": 2,
                "title_id": "20882",
                "url": "https://mangadex.org/title/{}",
                "url_format": "https://mangadex.org/chapter/{}",
              },
            ]
          }
          userFollows={Array []}
        />
      </WithStyles(ForwardRef(Grid))>
    </div>
    <WithStyles(ForwardRef(Tabs))
      onChange={[Function]}
      value={0}
    >
      <WithStyles(ForwardRef(Tab))
        label="Chapters"
        value={0}
      />
      <WithStyles(ForwardRef(Tab))
        label="Stats"
        value={1}
      />
    </WithStyles(ForwardRef(Tabs))>
    <TabPanelCustom
      index={0}
      noRerenderOnChange={true}
      value={0}
    >
      <ChapterList
        chapters={
          Array [
            Object {
              "chapter_number": 157,
              "chapter_url": "1007322",
              "group": "Shueisha",
              "release_date": 2020-07-05T16:00:00.000Z,
              "service_id": 1,
              "title": "Z=157: Same Time, Same Place",
              "url": "https://mangaplus.shueisha.co.jp/viewer/1007322",
            },
            Object {
              "chapter_number": 156,
              "chapter_url": "938629",
              "group": "MangaPlus",
              "release_date": 1970-01-01T00:00:00.000Z,
              "service_id": 2,
              "title": "Z=156: Two Scientists",
              "url": "https://mangadex.org/chapter/938629",
            },
            Object {
              "chapter_number": 156,
              "chapter_url": "1007024",
              "group": "Shueisha",
              "release_date": 2020-06-26T16:00:00.000Z,
              "service_id": 1,
              "title": "Z=156: Two Scientists",
              "url": "https://mangaplus.shueisha.co.jp/viewer/1007024",
            },
          ]
        }
        editable={false}
        mangaId={1}
        serviceUrlFormats={
          Object {
            "1": "https://mangaplus.shueisha.co.jp/viewer/{}",
            "2": "https://mangadex.org/chapter/{}",
          }
        }
      />
    </TabPanelCustom>
    <TabPanelCustom
      index={1}
      noRerenderOnChange={true}
      value={0}
    >
      <ReleaseHeatmap
        dataRows={Array []}
      />
    </TabPanelCustom>
  </WithStyles(ForwardRef(Paper))>
</WithStyles(ForwardRef(Container))>
`;

exports[`Manga page should render correctly should render correctly with little data 1`] = `
<WithStyles(ForwardRef(Container))
  disableGutters={true}
  maxWidth="lg"
>
  <WithStyles(ForwardRef(Paper))
    className="makeStyles-paper-10"
  >
    <div
      className="makeStyles-titleBar-2"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-title-1"
        variant="h4"
      >
        Dr. STONE
      </WithStyles(ForwardRef(Typography))>
    </div>
    <div
      className="makeStyles-details-5"
    >
      <a
        href={null}
        rel="noreferrer noopener"
        target="_blank"
      >
        <img
          alt="Dr. STONE"
          className="makeStyles-thumbnail-4"
          src={null}
        />
      </a>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-rootGrid-12"
        container={true}
        justify="space-between"
      >
        <WithStyles(ForwardRef(Grid))
          className="makeStyles-infoGrid-11"
          container={true}
          direction="column"
        >
          <MangaInfo
            mangaData={
              Object {
                "al": "https://anilist.co/manga/98416",
                "amz": "https://www.amazon.co.jp/gp/product/B075F8JBQ1",
                "ap": "https://www.anime-planet.com/manga/dr-stone",
                "artist": "Boichi",
                "author": "Inagaki Riichiro",
                "bw": "https://bookwalker.jp/series/114645",
                "chapters": Array [],
                "cover": null,
                "ebj": "https://www.ebookjapan.jp/ebj/413780/",
                "engtl": "https://www.viz.com/dr-stone",
                "estimated_release": null,
                "kt": "https://kitsu.io/manga/38860",
                "last_updated": null,
                "latest_chapter": null,
                "latest_release": null,
                "mal": null,
                "manga_id": 2,
                "mu": "https://www.mangaupdates.com/series.html?id=139601",
                "nu": "https://www.novelupdates.com/series/null",
                "raw": "null",
                "release_interval": undefined,
                "services": Array [
                  Object {
                    "name": "MANGA Plus",
                    "service_id": 1,
                    "title_id": "100010",
                    "url": "https://mangaplus.shueisha.co.jp/titles/{}",
                    "url_format": "https://mangaplus.shueisha.co.jp/viewer/{}",
                  },
                ],
                "status": 0,
                "title": "Dr. STONE",
              }
            }
          />
          <MangaAliases />
        </WithStyles(ForwardRef(Grid))>
        <MangaSourceList
          classesProp={
            Array [
              "makeStyles-sourceList-8",
            ]
          }
          followUnfollow={[Function]}
          items={
            Array [
              Object {
                "name": "MANGA Plus",
                "service_id": 1,
                "title_id": "100010",
                "url": "https://mangaplus.shueisha.co.jp/titles/{}",
                "url_format": "https://mangaplus.shueisha.co.jp/viewer/{}",
              },
            ]
          }
          userFollows={Array []}
        />
      </WithStyles(ForwardRef(Grid))>
    </div>
    <WithStyles(ForwardRef(Tabs))
      onChange={[Function]}
      value={0}
    >
      <WithStyles(ForwardRef(Tab))
        label="Chapters"
        value={0}
      />
      <WithStyles(ForwardRef(Tab))
        label="Stats"
        value={1}
      />
    </WithStyles(ForwardRef(Tabs))>
    <TabPanelCustom
      index={0}
      noRerenderOnChange={true}
      value={0}
    >
      <ChapterList
        chapters={Array []}
        editable={false}
        mangaId={2}
        serviceUrlFormats={
          Object {
            "1": "https://mangaplus.shueisha.co.jp/viewer/{}",
          }
        }
      />
    </TabPanelCustom>
    <TabPanelCustom
      index={1}
      noRerenderOnChange={true}
      value={0}
    >
      <ReleaseHeatmap
        dataRows={Array []}
      />
    </TabPanelCustom>
  </WithStyles(ForwardRef(Paper))>
</WithStyles(ForwardRef(Container))>
`;
